### PR TITLE
Add more logs, refactor metrics generation into one fucntion, placeholder methods generating model settings

### PR DESCRIPTION
**Pull request by KantiCodes**

### CODE REVIEW

**Comment by KantiCodes** [comment]
Piotrek obczaj to

File: src/model/data_loading.py [hunk 1]
```diff
@@ -150,15 +109,18 @@ def create_loaders(train_set, test_set, val_set, batch_size):
...
     return train_loader, test_loader, val_loader
 
 
-def dataset_metrics(dataset):
+def dataset_metrics(dataset, train_instances: list[Path]):
     total_positives = 0
     total_negatives = 0
     total_samples = 0
-    for d in dataset:
+    for d, file_name in zip(dataset, train_instances):
```
Comment (KantiCodes) [inline src/model/data_loading.py:123]: to dla debugowaia tylko

File: src/model/data_loading.py [hunk 2]
```diff
@@ -150,15 +109,18 @@ def create_loaders(train_set, test_set, val_set, batch_size):
     if val_set != []:
         val_loader = DataLoader(val_set, batch_size=len(val_set), shuffle=True)
 
+    one_batch_test = next(iter(test_loader))
+    print(f"Test set size of loader: {len(one_batch_test)}")
```
Comment (KantiCodes) [inline src/model/data_loading.py:113]: only for debugging

File: src/model/metrics.py [hunk 3]
```diff
@@ -29,11 +29,39 @@ def to_function(self):
...
+    precision_false: float = None
+    precision_true: float = None
+    recall_false: float = None
+    recall_true: float = None
+    f1_score_false: float = None
+    f1_score_true: float = None
+    orginal_number_of_false: int = None
+    orignal_number_of_true: int = None
+    predicted_number_of_false: int = None
+    predicted_number_of_true: int = None
```
Comment (KantiCodes) [inline src/model/metrics.py:42]: More info about metrics, in case we wanted to see how models behave

File: src/model/model_handler.py [hunk 4]
```diff
@@ -75,34 +72,30 @@ def load_model(self, model_path: str) -> None:
...
+            results.loss.backward()
             self.optimizer.step()
-        # loss, pred, original
-        return Results(loss, metric_result, out["operator"], batch["operator"].y)
-
-    def get_action_predictions(model: torch.nn.Module, state: torch.Tensor) -> torch.Tensor:
-        pass
+        
+        epoch_mean_results = Results.reduce_list_of_results(batch_results_list)
+        return epoch_mean_results
```
Comment (KantiCodes) [inline src/model/model_handler.py]: Previously we were reporting the results of last batch as epoch results, now for each epoch we get the average of all metrics. The `Results.reduce_lists_of_results` is a convenience method to get the average of all results

File: src/model/model_handler.py [hunk 5]
```diff
@@ -113,10 +106,8 @@ def predict_threshold(self, actions_proba, threshold: float):
     def test(self, data_loader: torch.utils.data.DataLoader) -> Results:
         self.model.eval()
         test_batch = next(iter(data_loader))
-        # if not all from plan are good
-        # then we have some key-map for the files
-        # add the test failed that failed xxx times more to the train set
```
Comment (KantiCodes) [inline src/model/model_handler.py:118]: useless comment

File: src/model/training.py [hunk 6]
```diff
@@ -105,13 +102,59 @@ def from_file(cls, path: str):
...
+                standardize_input_using_batch_norm=random.choice([True, False]),
+                conv_type_specific_kwargs={},  # You may need to specify values here based on the chosen conv_type
+                index=random.randint(0, 10),
+                loss_function=random.choice(list(ELossFunction)),
+                eval_metric=random.choice(list(EEvalMetric)),
+                batch_size=random.randint(8, 64)
+            )
+            settings_list.append(settings)
+        return settings_list
+
```
Comment (KantiCodes) [inline src/model/training.py]: Does not work, CHAT GPT sponsored, we can implement it later

File: src/model/training.py [hunk 7]
```diff
@@ -159,10 +200,10 @@ def get_model_handler(
     # We setup weights as global properties of the dataset only on the train set
     # to prevent overfitting
     # TODO hyperparameter on the weights
-    pos_weight, neg_weight = data_loading.calculate_weights(train_set)
+    pos_weight, neg_weight = data_loading.calculate_weights(train_set, train_instances)
 
     train_loader, test_loader, val_loader = data_loading.create_loaders(
-        train_set, test_set, val_set=val_set, batch_size=batch_size
+        train_set, test_set, val_set=val_set, batch_size=model_settings.batch_size
```
Comment (KantiCodes) [inline src/model/training.py:93]: Moved batch size to model_settings

File: src/model/training.py [hunk 8]
```diff
@@ -159,10 +200,10 @@ def get_model_handler(
     # We setup weights as global properties of the dataset only on the train set
     # to prevent overfitting
     # TODO hyperparameter on the weights
-    pos_weight, neg_weight = data_loading.calculate_weights(train_set)
+    pos_weight, neg_weight = data_loading.calculate_weights(train_set, train_instances)
```
Comment (KantiCodes) [inline src/model/training.py]: for debugging only

File: src/runner.py [hunk 9]
```diff
@@ -33,109 +33,19 @@ class Runner:
...
+        # TODO make this a parameter
+        all_instances = [
+            os.path.join(data_directory_path, x) for x in os.listdir(data_directory_path)
+        ]
+        # Shuffle the instances
+        random.shuffle(all_instances)
+        # Do random split here
+        self.training_instances = all_instances[: int(len(all_instances) * 0.8)]
+        self.test_instances = all_instances[int(len(all_instances) * 0.8) :]
+        self.val_instances = None
```
Comment (KantiCodes) [inline src/runner.py]: i am not saying it should live in the runner but the way it was before we had different `TestCases` running with different train/test dataset. This way the split is fixed

File: src/runner.py [hunk 10]
```diff
@@ -33,109 +33,19 @@ class Runner:
...
-    # #         test_instances=self.test_instances,
-    # #         model_settings=model_settings,
-    # #         num_epochs=30,
-    # #         batch_size=8,
-    # #     ) # TODO: Add parameters
-
-    # def persist(self):
-    #     # Save metrics and models
-    #     # MlFlow suff here
-    #     pass
```
Comment (KantiCodes) [inline src/runner.py:138]: Legacy code that we implemented the initial Runner with

File: src/runner.py [hunk 11]
```diff
@@ -151,10 +61,16 @@ def get_trainig_cases(self):
...
         for model_setting in os.listdir(self.directory_with_jsons):
             model_setting_path = os.path.join(self.directory_with_jsons, model_setting)
-            training_cases.append(TrainingCase(model_settings_path=model_setting_path))
+            training_cases.append(
+                TrainingCase(
+                    model_settings_path=model_setting_path,
+                    training_instances=self.training_instances,
+                    test_instances=self.test_instances,
+                    val_instances=self.val_instances)
+            )
```
Comment (KantiCodes) [inline src/runner.py]: See comment above about split between train/test

File: src/runner.py [hunk 12]
```diff
@@ -151,10 +61,16 @@ def get_trainig_cases(self):
...
         return training_cases
 
 
 if __name__ == "__main__":
-    runner_all_settings = Runner(directory_with_jsons="model-settings")
+    runner_all_settings = Runner(directory_with_jsons="model-settings", data_directory_path="data/preprocessed/blocksworld_graph/training/easy")
```
Comment (KantiCodes) [inline src/runner.py]: See comment about train/test split

File: src/training_case.py [hunk 13]
```diff
@@ -18,28 +20,20 @@ class TrainingCase:
     batch_size: int
     result_dict: dict
 
-    def __init__(self, model_settings_path: str):
+    def __init__(self, model_settings_path: str, training_instances: list, test_instances: list, val_instances: list = None):
```
Comment (KantiCodes) [inline src/training_case.py]: see above comment about train/test split

**Review by Czubbi — APPROVED** [review]

**Pull request merged**